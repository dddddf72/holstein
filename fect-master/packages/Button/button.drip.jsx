import { defineComponent, computed, ref, toRefs, onMounted } from 'vue'
import './button.drip.less'

const ButtonDrip = defineComponent({
  props: {
    x: [String, Number],
    y: [String, Number],
    onCompleted: Function,
  },
  setup(props) {
    const dripRef = ref(null)
    const { x, y, onCompleted } = toRefs(props)

    const calcPlace = computed(() => {
      const top = Number.isNaN(+y.value) ? 0 : y.value - 10
      const left = Number.isNaN(+x.value) ? 0 : x.value - 10
      return { top, left }
    })
    onMounted(() => {
      if (!dripRef.value) return
      dripRef.value.addEventListener('animationend', onCompleted.value())
      dripRef.value.removeEventListener('animationend', onCompleted.value())
    })

    return () => (
      <div ref={dripRef} className={'fect-button-drip'}>
        <svg width="20" height="20" viewBox="0 0 20 20" style={calcPlace.value}>
          <g stroke="none" strokeWidth="1" fill="none" fillRule="evenodd">
            <g fill="var(--accents-2)">
              <rect width="100%" height="100%" rx="10" />
            </g>
          </g>
        </svg>
      </div>
    )
  },
})

export default ButtonDrip
