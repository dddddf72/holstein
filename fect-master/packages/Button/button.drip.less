.fect-button-drip {
  position: absolute;
  left: 0;
  top: 0;
  right: 0;
  bottom: 0;
  // background-color: #ddd;
  // background-color: red;
  svg {
    position: absolute;
    animation: 350ms ease-in expand;
    animation-fill-mode: forwards;
    width: 16px;
    height: 16px;
  }
  @keyframes expand {
    0% {
      opacity: 0;
      transform: scale(1);
    }
    30% {
      opacity: 1;
    }
    80% {
      opacity: 0.5;
    }
    100% {
      transform: scale(28);
      opacity: 0;
    }
  }
}
