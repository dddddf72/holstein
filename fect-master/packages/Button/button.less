.fect-btn {
  box-sizing: border-box;
  display: inline-block;
  font-weight: 400;
  user-select: none;
  text-transform: capitalize;
  white-space: nowrap;
  text-align: center;
  outline: none;
  width: auto;
  height: 40px;
  border-radius: 5px;
  font-size: 15px;

  color: var(--accents-5);
  padding: 0 25px;
  overflow: hidden;
  cursor: pointer;
  position: relative;
  transition: border 0.2s, background 0.2s, color 0.2s ease-out;
  &:not(.auto) {
    min-width: 200px;
  }

  &.mini {
    width: initial;
    min-width: 150px;
    font-size: 12px;
    padding: 0 15px;
    height: 30px;
  }

  &.small {
    width: initial;
    min-width: 180px;
    font-size: 13px;
    padding: 0 20px;
    height: 35px;
  }
  &.large {
    width: initial;
    min-width: 250px;
    font-size: 18px;
    padding: 0 30px;
    height: 45px;
  }
  &.disabled {
    color: #ccc !important;
    cursor: not-allowed !important;
    // pointer-events: none;
    background: var(--accents-1) !important;
    border-color: var(--accents-2) !important;
    &:hover {
      background: var(--accents-1) !important;
      border-color: var(--accents-2) !important;
      color: #ccc !important;
      cursor: not-allowed !important;
    }
  }
  &.shadow {
    box-shadow: 0 3px 10px rgba(0, 0, 0, 0.22);
    &:not(.loading):hover {
      box-shadow: 0 7px 20px rgba(0, 0, 0, 0.2);
      transform: translateY(-1px);
    }
  }
  &.loading {
    // border: 1px solid #333 !important;
    // background-color: black;
    pointer-events: none;
    &.fect-loading-icon {
      background-color: #000;
    }
  }
  &.auto {
    min-width: auto;
  }
}

.fect-btn {
  background-color: var(--primary-background);
  border: 1px solid var(--accents-2);
  &.default {
    background-color: var(--primary-background);
    border: 1px solid var(--accents-2);
    &:not(.shadow):not(.loading):hover {
      color: var(--accents-1);
      background-color: var(--primary-background);
      color: #333;
      border-color: var(--primary-foreground);
    }
  }
  &.success {
    border-color: var(--success-default);
    background-color: var(--success-default);
    color: #fff;
    &:hover {
      color: var(--success-default);
      background-color: var(--primary-background);
    }
  }
  &.warning {
    border-color: var(--warning-default);
    background-color: var(--warning-default);
    color: #fff;
    &:hover {
      color: var(--warning-default);
      background-color: var(--primary-background);
    }
  }
  &.error {
    background-color: var(--error-default);
    border-color: var(--error-default);
    color: #fff;
    &:hover {
      color: var(--error-default);
      background-color: var(--primary-background);
    }
  }
}
