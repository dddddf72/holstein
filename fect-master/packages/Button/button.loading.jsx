import { defineComponent } from 'vue'

import Loading from '../Loading'

import './button.loading.less'

const ButtonLoading = defineComponent({
  setup(props, { attrs }) {
    return () => (
      <div className={`fect-btn-loading ${attrs.color}`}>
        <Loading />
      </div>
    )
  },
})

export default ButtonLoading
