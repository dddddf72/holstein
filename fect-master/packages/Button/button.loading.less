.fect-btn-loading {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 2;
  border-color: var(--accents-1);
  background-color: var(--accents-1);
  &.success {
    border-color: var(--success-default);
    background-color: var(--success-default);
    color: #fff;
  }
  &.warning {
    border-color: var(--warning-default);
    background-color: var(--warning-default);
    color: #fff;
  }
  &.error {
    background-color: var(--error-default);
    border-color: var(--error-default);
    color: #fff;
  }
}
