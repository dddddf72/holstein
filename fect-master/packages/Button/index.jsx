import { computed, ref, toRefs, watchEffect } from 'vue'
import { validator, theme, createNameSpace } from '../utils'
const { buttonTypes, normalSizes } = theme
const [createComponent] = createNameSpace('Button')

import ButtonDrip from './button.drip'
import ButtonLoading from './button.loading'
import './button.less'

export default createComponent({
  props: {
    type: {
      type: String,
      validator: validator.enums(buttonTypes),
      default: 'default',
    },
    size: {
      type: String,
      validator: validator.enums(normalSizes),
      default: 'medium',
    },
    disabled: Boolean,
    shadow: Boolean,
    loading: Boolean,
    auto: Boolean,
  },
  emits: ['click'],
  setup(props, { attrs, slots, emit }) {
    const { size, type, loading, shadow, disabled, auto } = toRefs(props)
    const buttonRef = ref(null)
    const drapShow = ref(false) // control drap component display
    const drapX = ref(0)
    const drapY = ref(0)

    const calcClass = computed(() => {
      let str = ''
      loading.value && (str += ' loading')
      type.value && (str += ` ${type.value}`)
      size.value && (str += ` ${size.value}`)
      shadow.value && (str += ' shadow')
      disabled.value && (str += ' disabled')
      auto.value && (str += ' auto')
      return str.trim()
    })

    const clickHandler = (e) => {
      // hide drip when button in shadow status
      if (disabled.value || loading.value) return
      const showDrip = !shadow.value
      if (showDrip) {
        const rect = buttonRef.value.getBoundingClientRect()
        drapShow.value = true
        drapX.value = e.clientX - rect.left
        drapY.value = e.clientY - rect.top
      }
      emit('click', e)
    }

    const dripCompleteHandler = () => {
      setTimeout(() => {
        drapShow.value = false
        drapX.value = 0
        drapY.value = 0
      }, 500)
    }

    return () => (
      <>
        <button
          {...attrs}
          disabled={disabled.value}
          className={`fect-btn ${calcClass.value} ${
            attrs.class ? attrs.class : ''
          }`}
          ref={buttonRef}
          onClick={clickHandler}
        >
          {loading.value && <ButtonLoading color={type.value} />}
          {drapShow.value && (
            <ButtonDrip
              x={drapX.value}
              y={drapY.value}
              onCompleted={dripCompleteHandler}
            />
          )}
          {slots.default?.()}
        </button>
      </>
    )
  },
})
