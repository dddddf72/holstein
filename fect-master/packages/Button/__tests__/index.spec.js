import { mount } from '@vue/test-utils'
import Button from '../index'
import { ref } from 'vue'

describe('Button', () => {
  it('should be render as Element', () => {
    const button = mount(Button)
    expect(button.html()).toMatchSnapshot()
  })

  it('should emit click event', () => {
    const wrapper = mount(Button)
    const button = wrapper.find('button')
    button.trigger('click')
    expect(wrapper.emitted()).toHaveProperty('click')
    // expect(button.emitted('click').length).toEqual(1)
  })

  it('should no emit click event when status as disabled', () => {
    const wrapper = mount(Button, {
      props: {
        disabled: true,
      },
    })
    expect(wrapper.emitted()).toEqual({})
  })

  it('should no emit click event when sstatus as loading', () => {
    const wrapper = mount(Button, {
      props: {
        loading: true,
      },
    })
    expect(wrapper.emitted()).toEqual({})
  })

  it('should allowed all types', () => {
    const wrapper = mount({
      setup(props) {
        return {
          typeList: ['default', 'success', 'warning', 'error'],
        }
      },
      render(h) {
        return (
          <>
            {this.typeList.map((item, i) => (
              <Button type={item} key={i} />
            ))}
          </>
        )
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
    expect(() => wrapper.unmount()).not.toThrow()
  })

  it('should allowed all size', () => {
    const wrapper = mount({
      setup(props) {
        return {
          sizeList: ['mini', 'small', 'medium', 'large'],
        }
      },
      render(h) {
        return (
          <>
            {this.sizeList.map((item, i) => (
              <Button size={item} key={i} />
            ))}
          </>
        )
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(() => wrapper.unmount()).not.toThrow()
  })

  it('should be support default render', async () => {
    const wrapper = mount({
      setup() {
        const val = ref(0)
        const handlerClick = () => (val.value = 1)
        return { val, handlerClick }
      },
      render() {
        return <Button onClick={this.handlerClick}>{this.val}</Button>
      },
    })

    expect(wrapper.find('.fect-btn').text()).toBe('0')
    await wrapper.find('.fect-btn').trigger('click')
    expect(wrapper.find('.fect-btn').text()).toBe('1')
  })
})
