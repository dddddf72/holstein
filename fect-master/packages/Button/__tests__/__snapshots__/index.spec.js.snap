// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should allowed all size 1`] = `
<button class="fect-btn default mini ">
  <!---->
  <!---->
  <!---->
</button><button class="fect-btn default small ">
  <!---->
  <!---->
  <!---->
</button><button class="fect-btn default medium ">
  <!---->
  <!---->
  <!---->
</button><button class="fect-btn default large ">
  <!---->
  <!---->
  <!---->
</button>
`;

exports[`Button should allowed all types 1`] = `
<button class="fect-btn default medium ">
  <!---->
  <!---->
  <!---->
</button><button class="fect-btn success medium ">
  <!---->
  <!---->
  <!---->
</button><button class="fect-btn warning medium ">
  <!---->
  <!---->
  <!---->
</button><button class="fect-btn error medium ">
  <!---->
  <!---->
  <!---->
</button>
`;

exports[`Button should be render as Element 1`] = `
<button class="fect-btn default medium ">
  <!---->
  <!---->
  <!---->
</button>
`;
